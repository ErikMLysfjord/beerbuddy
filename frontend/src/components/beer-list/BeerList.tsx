import InfiniteScroll from "react-infinite-scroll-component";
import BeerCard from "../beer-card/BeerCard";
import styles from "./BeerList.module.css";
import { useEffect, useRef } from "react";
import { useContext } from "react";
import { FilterContext } from "../../context/FilterContext";

type ReactionType = "unreact" | "upvote" | "downvote";

interface BeerListProps {
  beers: {
    beer_id: number;
    beer_name: string;
    brewery_name: string;
    vote_sum: number;
    beer_count: number;
    reaction: ReactionType;
  }[];
  fetchMore: (reset?: boolean, noFilters?: boolean) => Promise<void>;
}

/**
 * Helper function for translating the sorting method to a more readable format.
 * @param sorting - The sorting method to translate.
 * @returns - The translated sorting method.
 */
const translateSorting = (sorting: string) => {
  switch (sorting) {
    case "top":
      return "Most popular";
    case "low":
      return "Least popular";
    case "atoz":
      return "A-Z";
    case "ztoa":
      return "Z-A";
    default:
      return "Most popular";
  }
};

/**
 * A component for displaying a list of beers.
 * Uses lazy loading to load more beers when the user scrolls down.
 * @param props - The interface for the BeerList component.
 * @returns  - The beer list component.
 */
const BeerList = ({ beers, fetchMore }: BeerListProps) => {
  const { searchString, sorting } = useContext(FilterContext);

  const fetchMoreRef = useRef(fetchMore);
  fetchMoreRef.current = fetchMore;

  // Updates the beer list when the search string or sorting method changes.
  useEffect(() => {
    fetchMoreRef.current(true);
  }, [searchString, sorting]);

  return (
    <>
      <section
        className={styles.resultsInfoContainer}
        aria-label="Search and sorting information"
      >
        <div className={styles.resultsHeader}>
          <h2 className={styles.resultsInfo}>
            {beers[0]?.beer_count ?? 0} results
          </h2>
        </div>
        <p className={styles.resultsInfo}>Searched for: "{searchString}"</p>
        <p className={styles.resultsInfo}>
          Sorted by: {translateSorting(sorting)}
        </p>
      </section>
      <section aria-label="Beer list">
        <InfiniteScroll
          dataLength={beers.length}
          next={fetchMoreRef.current}
          hasMore={beers?.length < beers[0]?.beer_count}
          loader={
            <p style={{ textAlign: "center" }}>
              <b>Loading...</b>
            </p>
          }
          endMessage={
            <p style={{ textAlign: "center" }}>
              <b>Yay! You have seen it all</b>
            </p>
          }
          scrollThreshold={0.99}
          scrollableTarget="infiniteScrollTarget"
        >
          <ul className={styles.list}>
            {beers?.map((beer) => (
              <li key={beer.beer_id}>
                <BeerCard
                  beer_id={beer.beer_id}
                  name={beer.beer_name}
                  brewery={beer.brewery_name}
                  votes={beer.vote_sum}
                  reaction={beer.reaction}
                />
              </li>
            ))}
          </ul>
        </InfiniteScroll>
      </section>
    </>
  );
};
export default BeerList;
