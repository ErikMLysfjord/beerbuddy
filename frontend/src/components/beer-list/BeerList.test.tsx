import { describe, it, expect, vi, afterEach } from "vitest";
import { render, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom";
import BeerList from "./BeerList";

describe("BeerList", () => {
  const mockData = {
    data: {
      beers: [
        {
          beer_name: "21st Amendment Bitter American",
          brewery_name: "21st Amendment Brewery Cafe",
          vote_sum: "0",
        },
        {
          beer_name: "Borg Citra",
          brewery_name: "Borg BrugghÃºs",
          vote_sum: "0",
        },
        {
          beer_name: "Sierra Nevada Pale Ale",
          brewery_name: "Sierra Nevada Brewing Company",
          vote_sum: "0",
        },
        {
          beer_name: "Mono Stereo Mosaic",
          brewery_name: "Mono Brewing Co.",
          vote_sum: "0",
        },
      ],
    },
  };

  global.fetch = vi.fn(() =>
    Promise.resolve(new Response(JSON.stringify(mockData)))
  );

  afterEach(() => {
    vi.clearAllMocks();
  });

  it("renders correctly", () => {
    const { container } = render(<BeerList />);
    expect(container).toMatchSnapshot();
  });

  it("renders with correct number of beers", async () => {
    const { getByText, getAllByRole } = render(<BeerList />);
    expect(getByText("Loading...")).toBeInTheDocument();

    await waitFor(() => {
      expect(getAllByRole("listitem")).toHaveLength(4);
    });
  });

  it("renders with correct names", async () => {
    const { getByText } = render(<BeerList />);
    expect(getByText("Loading...")).toBeInTheDocument();

    await waitFor(() => {
      expect(getByText("21st Amendment Bitter American")).toBeInTheDocument();
      expect(getByText("21st Amendment Bitter American")).toBeInstanceOf(
        HTMLHeadingElement
      );
      expect(getByText("Borg Citra")).toBeInTheDocument();
      expect(getByText("Sierra Nevada Pale Ale")).toBeInTheDocument();
      expect(getByText("Mono Stereo Mosaic")).toBeInTheDocument();
    });
  });
});
