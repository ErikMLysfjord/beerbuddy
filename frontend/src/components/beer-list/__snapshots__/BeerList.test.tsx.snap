// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BeerList > renders correctly 1`] = `
<div>
  <section
    class="_resultsInfoContainer_ac700a"
  >
    <div
      class="_resultsHeader_ac700a"
    >
      <h2
        class="_resultsInfo_ac700a"
      >
        100
         results
      </h2>
      <button
        class="ant-btn css-dev-only-do-not-override-pr0fja ant-btn-primary"
        type="button"
      >
        <span>
          Reset filters
        </span>
      </button>
    </div>
    <p
      class="_resultsInfo_ac700a"
    >
      Searched for: "
      "
    </p>
    <p
      class="_resultsInfo_ac700a"
    >
      Sorted by: 
      Most popular
    </p>
  </section>
  <main>
    <div
      class="infinite-scroll-component__outerdiv"
    >
      <div
        class="infinite-scroll-component "
        style="height: auto; overflow: auto;"
      >
        <ul
          class="_list_ac700a"
        >
          <li>
            <a
              aria-label="21st Amendment Bitter American"
              class="_card_539764"
              href="./beer/1"
            >
              <div
                class="_textWrapper_539764"
              >
                <h1
                  class="_beerName_539764"
                >
                  21st Amendment Bitter American
                </h1>
                <p
                  class="_breweryName_539764"
                >
                  21st Amendment Brewery Cafe
                </p>
              </div>
              <div
                class="_wrapper_0f0ab2"
              >
                <button
                  aria-label="Upvote this beer"
                  class="ant-btn css-dev-only-do-not-override-pr0fja ant-btn-primary ant-btn-icon-only"
                  style="filter: brightness(0.8);"
                  type="button"
                >
                  <span
                    class="ant-btn-icon"
                  >
                    <span
                      aria-label="caret-up"
                      class="anticon anticon-caret-up"
                      role="img"
                    >
                      <svg
                        aria-hidden="true"
                        data-icon="caret-up"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="0 0 1024 1024"
                        width="1em"
                      >
                        <path
                          d="M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z"
                        />
                      </svg>
                    </span>
                  </span>
                </button>
                <p
                  class="_count_0f0ab2"
                >
                  NaN
                </p>
                <button
                  aria-label="Downvote this beer"
                  class="ant-btn css-dev-only-do-not-override-pr0fja ant-btn-primary ant-btn-icon-only"
                  style="filter: brightness(0.8);"
                  type="button"
                >
                  <span
                    class="ant-btn-icon"
                  >
                    <span
                      aria-label="caret-down"
                      class="anticon anticon-caret-down"
                      role="img"
                    >
                      <svg
                        aria-hidden="true"
                        data-icon="caret-down"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="0 0 1024 1024"
                        width="1em"
                      >
                        <path
                          d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                        />
                      </svg>
                    </span>
                  </span>
                </button>
              </div>
            </a>
          </li>
          <li>
            <a
              aria-label="Borg Citra"
              class="_card_539764"
              href="./beer/2"
            >
              <div
                class="_textWrapper_539764"
              >
                <h1
                  class="_beerName_539764"
                >
                  Borg Citra
                </h1>
                <p
                  class="_breweryName_539764"
                >
                  Borg Bruggh√∫s
                </p>
              </div>
              <div
                class="_wrapper_0f0ab2"
              >
                <button
                  aria-label="Upvote this beer"
                  class="ant-btn css-dev-only-do-not-override-pr0fja ant-btn-primary ant-btn-icon-only"
                  style="filter: brightness(0.8);"
                  type="button"
                >
                  <span
                    class="ant-btn-icon"
                  >
                    <span
                      aria-label="caret-up"
                      class="anticon anticon-caret-up"
                      role="img"
                    >
                      <svg
                        aria-hidden="true"
                        data-icon="caret-up"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="0 0 1024 1024"
                        width="1em"
                      >
                        <path
                          d="M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z"
                        />
                      </svg>
                    </span>
                  </span>
                </button>
                <p
                  class="_count_0f0ab2"
                >
                  NaN
                </p>
                <button
                  aria-label="Downvote this beer"
                  class="ant-btn css-dev-only-do-not-override-pr0fja ant-btn-primary ant-btn-icon-only"
                  style="filter: brightness(0.8);"
                  type="button"
                >
                  <span
                    class="ant-btn-icon"
                  >
                    <span
                      aria-label="caret-down"
                      class="anticon anticon-caret-down"
                      role="img"
                    >
                      <svg
                        aria-hidden="true"
                        data-icon="caret-down"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="0 0 1024 1024"
                        width="1em"
                      >
                        <path
                          d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                        />
                      </svg>
                    </span>
                  </span>
                </button>
              </div>
            </a>
          </li>
          <li>
            <a
              aria-label="Sierra Nevada Pale Ale"
              class="_card_539764"
              href="./beer/3"
            >
              <div
                class="_textWrapper_539764"
              >
                <h1
                  class="_beerName_539764"
                >
                  Sierra Nevada Pale Ale
                </h1>
                <p
                  class="_breweryName_539764"
                >
                  Sierra Nevada Brewing Company
                </p>
              </div>
              <div
                class="_wrapper_0f0ab2"
              >
                <button
                  aria-label="Upvote this beer"
                  class="ant-btn css-dev-only-do-not-override-pr0fja ant-btn-primary ant-btn-icon-only"
                  style="filter: brightness(0.8);"
                  type="button"
                >
                  <span
                    class="ant-btn-icon"
                  >
                    <span
                      aria-label="caret-up"
                      class="anticon anticon-caret-up"
                      role="img"
                    >
                      <svg
                        aria-hidden="true"
                        data-icon="caret-up"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="0 0 1024 1024"
                        width="1em"
                      >
                        <path
                          d="M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z"
                        />
                      </svg>
                    </span>
                  </span>
                </button>
                <p
                  class="_count_0f0ab2"
                >
                  NaN
                </p>
                <button
                  aria-label="Downvote this beer"
                  class="ant-btn css-dev-only-do-not-override-pr0fja ant-btn-primary ant-btn-icon-only"
                  style="filter: brightness(0.8);"
                  type="button"
                >
                  <span
                    class="ant-btn-icon"
                  >
                    <span
                      aria-label="caret-down"
                      class="anticon anticon-caret-down"
                      role="img"
                    >
                      <svg
                        aria-hidden="true"
                        data-icon="caret-down"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="0 0 1024 1024"
                        width="1em"
                      >
                        <path
                          d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                        />
                      </svg>
                    </span>
                  </span>
                </button>
              </div>
            </a>
          </li>
          <li>
            <a
              aria-label="Mono Stereo Mosaic"
              class="_card_539764"
              href="./beer/4"
            >
              <div
                class="_textWrapper_539764"
              >
                <h1
                  class="_beerName_539764"
                >
                  Mono Stereo Mosaic
                </h1>
                <p
                  class="_breweryName_539764"
                >
                  Mono Brewing Co.
                </p>
              </div>
              <div
                class="_wrapper_0f0ab2"
              >
                <button
                  aria-label="Upvote this beer"
                  class="ant-btn css-dev-only-do-not-override-pr0fja ant-btn-primary ant-btn-icon-only"
                  style="filter: brightness(0.8);"
                  type="button"
                >
                  <span
                    class="ant-btn-icon"
                  >
                    <span
                      aria-label="caret-up"
                      class="anticon anticon-caret-up"
                      role="img"
                    >
                      <svg
                        aria-hidden="true"
                        data-icon="caret-up"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="0 0 1024 1024"
                        width="1em"
                      >
                        <path
                          d="M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z"
                        />
                      </svg>
                    </span>
                  </span>
                </button>
                <p
                  class="_count_0f0ab2"
                >
                  NaN
                </p>
                <button
                  aria-label="Downvote this beer"
                  class="ant-btn css-dev-only-do-not-override-pr0fja ant-btn-primary ant-btn-icon-only"
                  style="filter: brightness(0.8);"
                  type="button"
                >
                  <span
                    class="ant-btn-icon"
                  >
                    <span
                      aria-label="caret-down"
                      class="anticon anticon-caret-down"
                      role="img"
                    >
                      <svg
                        aria-hidden="true"
                        data-icon="caret-down"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="0 0 1024 1024"
                        width="1em"
                      >
                        <path
                          d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                        />
                      </svg>
                    </span>
                  </span>
                </button>
              </div>
            </a>
          </li>
        </ul>
        <p
          style="text-align: center;"
        >
          <b>
            Loading...
          </b>
        </p>
      </div>
    </div>
  </main>
</div>
`;
